feat: add `auth` field to `RPCEndpointConfig`

caiquejjx opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.
-->

## Motivation
Closes #8276 

<!--
Explain the context and why you're making that change. What is the problem
you're trying to solve? In some cases there is not a problem and this can be
thought of as being the motivation for your change.
-->

## Solution
Adds the auth field to RPCEndpointConfig

zerosnacks commented on crates/config/src/lib.rs:
```diff
@@ -886,7 +886,7 @@ impl Config {
                         if self.offline {
                             return Err(SolcError::msg(format!(
                                 "can't install missing solc {version} in offline mode"
-                            )))
```
Nit: please undo the inserted `;`

See: https://github.com/foundry-rs/foundry/blob/master/CONTRIBUTING.md#resolving-an-issue for the linter settings (we are using `nightly` fmt and `nightly` clippy)

zerosnacks commented:
Thanks for your PR! Small note on the unintended formatting diff

caiquejjx commented:
done, thanks @zerosnacks

mattsse reviewed (changes requested):
great start,

we can expect that this will be used via some env var or similar so we should add support for resolving env vars

see for ref

https://github.com/foundry-rs/foundry/blob/c99854277c346fa6de7a8f9837230b36fd85850e/crates/config/src/endpoints.rs#L117-L131

and

https://github.com/foundry-rs/foundry/blob/c99854277c346fa6de7a8f9837230b36fd85850e/crates/config/src/endpoints.rs#L150-L160

caiquejjx commented:
@mattsse I have implemented the auth as an enum similar to the `RpcEndpoint`, lmk you thoughts

mattsse commented on crates/config/src/endpoints.rs:
```diff
@@ -210,6 +210,58 @@ impl From<RpcEndpoint> for RpcEndpointConfig {
+/// It works in the same way as the `RpcEndpoint` type, where it can be a raw string or a reference
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum RpcAuth {
+    Token(String),
```
```suggestion
    Raw(String),
```

mattsse reviewed (changes requested):
nice, would love to see some additional tests like

https://github.com/foundry-rs/foundry/blob/cce36f85c80946bd4a1868411aa99eda879a0e43/crates/config/src/lib.rs#L3369-L3369

caiquejjx commented:
@mattsse I added a test for checking the resolving of the auth and I also added the `auths` field to the `ResolvedRpcEndpoint` as a way of resolving the auth, its probably not the most optimal solution but is the one that requires less changes on existing code. I can also create a new type for the `config.resolve` result where I can add the auth for an endpoint but let me know what you think

mattsse reviewed (approved):
cool, nice work.

next up we'd find a way to integrate this properly

I think the entrypoint for this would be:

https://github.com/foundry-rs/foundry/blob/master/crates/cli/src/utils/mod.rs#L92-L92

mattsse commented:
failing tests unrelated
